### PR TITLE
Implement search lexer for breaking up search

### PR DESCRIPTION
**Pull request by mattdean-digicatapult**
Implements a search lexer for breaking up searches into different types of field lookup.

What's implemented:
1. custom field names
2. breaking up on search into terms and field matches
3. quoting to combine words into a single term
4. modifiers such as `+` and `-` so we can implement strong match and negative match
5. validation of field names to `/^[a-zA-Z]$/` so we can later implement comparitives

What's not implemented:
1. Any sort of UI indication this exists. It is hidden magic
2. Input validation for things like numbers, currency values and dates. These will have to come with a ui update to indicate sections of a search that are invalid and why
3. A form to simplify the creation of complex queries
4. Comparative queries other than matches. For example `field_greater` or `field_less`

### CODE REVIEW

File: src/Search/lexer.spec.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+  test('reorder terms and fields', () => {
+    const result = parseSearch(defaultLexer, 'hello test:me world')
+    expect(result).toEqual([
+      { type: 'term', value: 'hello', modifier: null, isQuoted: false },
+      { type: 'fieldMatch', field: 'test', value: 'me', modifier: null, isQuoted: false },
+      { type: 'term', value: 'world', modifier: null, isQuoted: false }
+    ])
+  })
+
+  test('quoted match field', () => {
```
Comment (n3op2) [inline src/Search/lexer.spec.ts:70]: nice, i believe it was even quite comfortable to write tests for this feature

File: src/Search/lexer.ts [hunk 2]
```diff
@@ -0,0 +1,105 @@
+import moo from 'moo'
```
Comment (n3op2) [inline src/Search/lexer.ts:1]: need to checkout this module

**Review by jonmattgray — APPROVED** [review]

**Review by n3op2 — APPROVED** [review]
nice one!

File: src/Search/lexer.spec.ts (hunk 1 above)
Comment (mattdean-digicatapult) [reply src/Search/lexer.spec.ts:70]: it was alright. I should have done it a bit more tdd tbh

**Pull request merged**